include/real/exact_number.hpp

```diff
@@ -693,190 +694,322 @@ namespace boost {
+                if(divisor == zero){
+                    throw divide_by_zero();
+                }
 
-                // we ignore exponents, then set them in the end.
-                // (a * 10^x) / (b*10^y) = (a*10 / b*10) * 10^((x-1)-(y-1))
-                // 100/20 -> 1/2 * (10)
-                int exponent_dif = (numerator.exponent - 1) - (divisor.exponent - 1);
+                /* Special Cases */
+                if((*this) == zero){
+                    return;
+                }
 
-                numerator.exponent = 1;
-                divisor.exponent = 1;
-                tmp.digits = {1};
-                tmp.exponent = 1;
+                exact_number<T> one;
+                one.digits={1};
+                one.exponent=1;
+                if(divisor == one){
+                    return;
+                }
 
-                exact_number<T> zero = exact_number<T>(); 
+                bool positive = (this->positive == divisor.positive);
 
-                if (divisor == tmp) {
-                    this->exponent = exponent_dif + 1;
+                exact_number<T> minus_one = one;
+                minus_one.positive = false;
+                if(divisor == minus_one){
                     this->positive = positive;
                     return;
                 }
 
-                if (divisor == (*this)) { 
-                    (*this) = tmp;
-                    (*this).positive = positive;
+                if(divisor == (*this)){
+                    (*this) = one;
                     return;
                 }
 
-                if ((*this) == zero)
-                    return;
-                    
-                if (divisor == zero)
-                    throw(boost::real::divide_by_zero());
-
-                // N < D --> 0 < abs(Q) < 1
-                if (numerator < divisor) {
-                        left = exact_number<T>(); // 0
-                        right = tmp; // 1
-                    } else { // assuming D > 1. N > D ---> 1 < N / D < N
-                        left = tmp; // 1
-                        right = numerator;
-                    }
+                const T base = (std::numeric_limits<T>::max() / 4) * 2 - 1;
+                exact_number<T> half;
+                half.digits = {base / 2 + 1};
+
+                exact_number<T> left, right, numerator, denominator, residual, length;
+                numerator = (*this);
+                denominator = divisor;
+                numerator = numerator.abs();
+                denominator = denominator.abs();
 
-                // distance = (right - left) / 2
-                distance = (right - left) * half;
-                (*this) = left + distance;
-                // N/D = Q -> QD -N = 0
-                residual = (*this) * divisor - numerator;
-                if (residual == zero) {
-                    this->exponent += exponent_dif;
+                int exponent_diff = (this->exponent - 1) - (denominator.exponent - 1);
```

Comment:
Whats the 2 (-1) for? can you just remove them? 
   (a - 1) - (b -1)
= (a -b) -1 +1
= (a -b)

Reply:
Actually, a*base^(n) / b*base^(m)  =  ( a*base^(1) / b*base^(1) ) * ( base^(n-1) / base^(m-1) ) 
now ,  ( base^(n-1) / base^(m-1) )  = base ^ ( (n-1) - (m-1) )
So the (-1) was just for proper understanding. It's my fault that I didn't make proper comments. I'll add the comments.
Complexity wise, I don't think (-1) is an issue. Am I right?

Reply:
No, having the -1 only adds a constant extra time, but you can remove the -1 and add a good comment to write down the entire math involved in that line.